### ISSUE TITLE
Add a translator comment sniff

### ISSUE THREAD

**Issue by atimmer** [WordPress-Coding-Standards/WordPress-Coding-Standards#423]
All strings with placeholders should have a translator comment to explain what the placeholders are replaced with. It would be very nice to have a sniff to automatically flag strings without a proper translator comment.

These would be flagged as incorrect:

``` php

__( 'A %1$s to be translated', 'text domain' );

/* translators: %2$s is replaced with the number of translations */
__( 'A %1$s to be translated', 'text domain' );
```

And these would be fine:

``` php
__( 'A string to be translated', 'text domain' );

/* translators: %1$s is replaced with "string" */
__( 'A %1$s to be translated', 'text domain' );
```

Strings can be checked for placeholders using the following regex: `%(\d+\$(?:\d+)?)?[bcdefgosuxEFGX]`.

**Comment by westonruter** [comment]
:+1:

**Comment by johnbillion** [comment]
I'm not 100% sure, but I believe subsequent instances of an identical string should not be proceeded by the translator comment (in the same way subsequent uses of a hook or filter shouldn't be documented). Needs checking.

**Comment by atimmer** [comment]
@johnbillion This depends on whether or not the pot generation will correctly pickup the translator comment if it is only on one of the strings.

If that's the case there is still value in this sniff, there then needs to be a way to tell it that the translation has been documented elsewhere.

**Comment by JDGrimes** [comment]
Related: #567

**Comment by jrfnl** [comment]
All, I've started working on adding this check, but would like to verify the exact specs.

This is what I've found so far:
___
> It has to start with the words translators: and to be the last PHP comment before the gettext call.

Source: https://codex.wordpress.org/I18n_for_WordPress_Developers#Descriptions

> So that translators know how to translate a string like __( 'g:i:s a' ) you can add a clarifying comment in the source code. It has to start with the words translators: and to be the last PHP comment before the gettext call. Here is an example:
> ```php
> /* translators: draft saved date format, see http://php.net/date */
> $saved_date_format = __( 'g:i:s a' );
> ```
> It’s also used to explain placeholders in a string like _n_noop( '<strong>Version %1$s</strong> addressed %2$s bug.','<strong>Version %1$s</strong> addressed %2$s bugs.' ).
> ```php	
> /* translators: 1: WordPress version number, 2: plural number of bugs. */
> _n_noop( '<strong>Version %1$s</strong> addressed %2$s bug.',
>         '<strong>Version %1$s</strong> addressed %2$s bugs.' );
> ```
Source: https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/#descriptions (similar to: https://developer.wordpress.org/themes/functionality/internationalization/#descriptions)
___
However for a sniff those specs are a bit limited. So I have the following questions:
* What type of comments are acceptable ? Only single line star comments `/* */` as per the examples ? Or also `//` single line comments, multi-line star comments, doc blocks ?
```php
// translators:

/* transators: */

/*
 * translators:
 */

/**
 * translators:
*/
```

* _"must be the last PHP comment before the gettext call"_ This is open to interpretation - the last comment could be five lines above the translation call. In practice I've found that PoEdit will only pick up the translators comment if it's on the line right above the translators call. I'm not sure about other makepot tools. I'm inclined to make this quite a strict check where it is required for the comment to be on the line above the gettext call.
```php
printf(
	/* translators: number of monkeys, location. */
	__( 'There are %1$d monkeys in the %2$s', 'my-slug' ),
	intval( $number ),
	esc_html( $string )
); // This would be ok.

/* translators: number of monkeys, location. */
printf(
	__( 'There are %1$d monkeys in the %2$s', 'my-slug' ),
	intval( $number ),
	esc_html( $string )
); // This would NOT be ok.
```

Other than that I am making the following presumptions:
* Translators comments are **_required_** when any of the text strings (text, single, plural) contains a placeholder.
* If a translators comment is present, but there is no placeholder detected, do nothing. (i.e. the time format example above)

I will currently not address checking the number of placeholders against the translators comment as the formats used vary and this would IMHO be extremely prone to bugs. See the below examples to illustrate what I mean:
```php
/* translators: 1. number of monkeys, 2. location. */
/* translators: %1$d: number of monkeys, %2$s: location. */
/* translators: %d - number of monkeys, %s - location. */
/* translators: number of monkeys, location. */
__( 'There are %1$d monkeys in the %2$s', 'my-slug' );
```

**Comment by jrfnl** [comment]
Oh and input on whether or not some form of whitelisting for a gettext call without translators comment is needed and what form it should take would be welcome too!

**Comment by jrfnl** [comment]
And another question: error or warning ?

**Comment by GaryJones** [comment]
1. The default in core seems to be `/* translators: ... */`, though I know we have to disable a comment check for using a multiline comment format on a single line. `// translators: ` will work, and should be allowed too.

2. Agreed that it only seems to be the line immediately above (though I any blank lines between the comment and string may be ignored), so that's what we should allow. Poor wording in the spec.

3. There will be a lot of old code without translator comments, so I'd say Warning, and make it be recommended but not required (or required, under WordPress-Extra), if placeholders exist.

**Comment by jrfnl** [comment]
@GaryJones Good input. Thanks!

The questions regarding the documentation style and whether or not new lines are allowed is not just one of personal preference/"what does core do", but I think we also need to look at how tools to generate `.pot` files handle this.

I can think of a few:
* PoEdit
* makepot

But there are bound to be more tools out there and what about GlotPress ?

Anyone here who can give us some insight into the inner workings of these tools with regards to comment style and comment position ? or know of someone who does know ? If so, please ping them to alert them that their input would be much appreciated.

**Comment by GaryJones** [comment]
@ocean90 will know if GlotPress has any considerations here, or whether it internally uses makepot or similar.

**Comment by grappler** [comment]
Glotpress uses Makepot.

If I am not mistaken then the code that fetched the comments is in `find_function_calls ()` in https://i18n.svn.wordpress.org/tools/trunk/extract/extract.php which uses `T_COMMENT`

This core ticket maybe of interest too. https://core.trac.wordpress.org/ticket/39116

**Comment by jrfnl** [comment]
PR #742 is ready for testing ;-)
